contracts/Removal.sol

```diff
@@ -655,6 +643,72 @@ contract Removal is
     }
   }
 
+  /**
+   * @notice Hook that is called before before any token transfer. This includes minting and burning, as well as
+   * batched variants. Disables transfers to any address that is not the `Market` or `Certificate` contracts, the zero
+   * address (for burning), or the supplier address that is encoded in the token ID itself.
+   *
+   * @dev Follows the rules of hooks defined [here](
+   *  https://docs.openzeppelin.com/contracts/4.x/extending-contracts#rules_of_hooks)
+   *
+   * ##### Requirements:
+   *
+   * - The contract must not be paused.
+   * - Enforces the rules of `ERC1155Upgradeable._beforeTokenTransfer`.
+   * - Enforces the rules of `ERC1155SupplyUpgradeable._beforeTokenTransfer`.
+   */
+  function _beforeTokenTransfer(
+    address operator,
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  ) internal virtual override whenNotPaused {
+    address market = address(_market);
+    for (uint256 i = 0; i < ids.length; ++i) {
+      uint256 id = ids[i];
+      if (amounts[i] == 0) {
+        revert InvalidTokenTransfer({tokenId: id});
+      }
+      if (to == market) {
+        _currentMarketBalance += amounts[i];
+      }
+      if (from == market) {
+        _currentMarketBalance -= amounts[i];
+      }
+      if (
```

Comment:
This undoes our optimization from earlier

Reply:
oh jeez good catch, wonder how that happened